src/core_plugins/kibana/public/management/sections/indices/create_index_pattern/create_index_pattern.js

```diff
@@ -207,39 +195,54 @@ uiModules.get('apps/management')
     return loadingCount > 0;
   };
 
-  this.refreshFieldList = () => {
-    const timeField = this.formValues.timeField;
+  this.refreshTimeFieldOptions = () => {
+    const prevOption = this.formValues.timeFieldOption;
+
     loadingCount += 1;
-    fetchFieldList().then(results => {
-      if (timeField) {
-        updateFieldListAndSetTimeField(results, timeField.name);
-      } else {
-        updateFieldList(results);
-      }
-    }).finally(() => {
-      loadingCount -= 1;
-    });
+    this.timeFieldOptions = [];
+    this.timeFieldOptionsError = null;
+    this.formValues.timeFieldOption = null;
+    getTimeFieldOptions()
```

Comment:
I agree, we should avoid using $apply where possible. I think the way to do that is to follow a few simple rules:

1. Prefer Angular DOM event handlers like `ng-click` over custom event handlers (like `$(element).on('click', () => {})` and `element.addEventListener('click', () => {})`.'
2. Prefer Angular wrappers around global objects like `$document` and `$window`.
2. Prefer Angular wrappers around async actions like `$q`, `$timeout`, and `$interval`.

There are probably a few others we can dig up by Googling but the general theme is to just use whatever tools Angular makes available which will tie into the $digest cycle under the hood. These are situations where $apply is inappropriate.

This leaves us with the situations where $apply is inappropriate: anywhere these tools can't be used. I think if we can confidently assess a call stack and state that it will never touch any of these tools, and therefore never trigger a $digest cycle on its own, then we can safely use $apply. Thoughts @spalger and @chrisronline ?